### PR TITLE
Create community-nft1.0.sol

### PR DESCRIPTION
**Pull request by chy2021**
社区不少人和nft在用的合约，希望能帮优化到2.0

### CODE REVIEW

File: contracts/community-nft1.0.sol [hunk 1]
```diff
@@ -0,0 +1,297 @@
+pragma solidity ^0.6.2;
+pragma experimental ABIEncoderV2;
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:2]: If you use the latest solidity version, e.g. ^0.8.0, the ABIEncoderV2 is not experimental anymore.

File: contracts/community-nft1.0.sol [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+
+import "@openzeppelin/contracts/token/ERC1155/ERC1155.sol";
+import "@openzeppelin/contracts/utils/Counters.sol";
+import "../utils/StringUtils.sol";
+import "../SponsorWhitelistControl.sol";
+import "../owner/Operator.sol";
+import "@openzeppelin/contracts/utils/ReentrancyGuard.sol";
+
+contract TspaceAllnighter is ERC1155, Operator {
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:11]: This repo is common for all projects as OZ did, so please avoid any concrete project name.

On the other hand, I think the best place for such contract is `contracts/token/CRC1155/presets`, you could refer to the code structure of OZ.

You could refer to the OZ [implementation](https://github.com/OpenZeppelin/openzeppelin-contracts/blob/master/contracts/token/ERC1155/presets/ERC1155PresetMinterPauser.sol) for more details.

File: contracts/community-nft1.0.sol [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+    using Counters for Counters.Counter;
+    //tokenid自增一
+    Counters.Counter private _tokenIds;
+
+    using SafeMath for uint256;
+
+    // Contract name
+    string public name = "Tspace Allnighter";
+    // Contract symbol
+    string public symbol = "allnighter";
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:23]: Could inherit the existing `CRC1155Metadata`.

File: contracts/community-nft1.0.sol [hunk 4]
```diff
@@ -0,0 +1,297 @@
...
+
+    // Mapping from owner to list of owned token IDs
+    mapping(address => uint256[]) private _ownedTokens;
+
+    // Mapping from token ID to index of the owner tokens list
+    mapping(uint256 => uint256) private _ownedTokensIndex;
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:29]: Could reuse the existing `CRC1155Enumerable` to simplify code.

File: contracts/community-nft1.0.sol [hunk 5]
```diff
@@ -0,0 +1,297 @@
...
+
+    // Mapping from token id to position in the allTokens array
+    mapping(uint256 => uint256) private _allTokensIndex;
+
+    string baseUri;
+
+    mapping(uint256 => address[]) public ownerOfAddress;
+
+    //tokenId => metadate
+    mapping(uint256 => string) public tokenMetaDate;
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:42]: Per EIP-1155, you could use a common URI with placeholder for all tokens, which also reduce the gas/storage cost.

File: contracts/community-nft1.0.sol [hunk 6]
```diff
@@ -0,0 +1,297 @@
...
+
+    uint256 public price = 5e18;
+
+    SponsorWhitelistControl public constant SPONSOR =
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:46]: Please refer to the existing library `InternalContracts`.

File: contracts/community-nft1.0.sol [hunk 7]
```diff
@@ -0,0 +1,297 @@
...
+        );
+
+    //设置baseUri,用来获取中心化服务
+    constructor() public ERC1155("https://tspace-1253470014.cos.ap-hongkong.myqcloud.com/nft/allnighter/metadata/") {
+        baseUri = "https://tspace-1253470014.cos.ap-hongkong.myqcloud.com/nft/allnighter/metadata/";
+
+        //register all users as sponsees
+        address[] memory users = new address[](1);
+        users[0] = address(0);
+        SPONSOR.addPrivilege(users);
```
Comment (boqiu) [inline contracts/community-nft1.0.sol:58]: You could use existing `InternalContractsHandler` to simplify your code.

**Review by boqiu — CHANGES_REQUESTED** [review]

File: contracts/community-nft1.0.sol [hunk 8]
```diff
@@ -0,0 +1,297 @@
...
+
+    function tokensOf(address _owner)
+        public
+        view
+        returns (uint256[] memory _tokens)
+    {
+        return _ownedTokens[_owner];
+    }
+
+    //根据tokenId 获取 metadate
```
Comment (Pana) [inline contracts/community-nft1.0.sol:93]: Chinese is not recommended in code.

**Review by Pana — CHANGES_REQUESTED** [review]

**Pull request closed without merging**